sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/ObjectPool.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+import java.util.function.Supplier;
+
+/**
+ * A pool of objects of type {@code T}.
+ *
+ * <p>When an object is borrowed from an empty pool, an object will be created by the supplied
+ * {@code objectCreator} and returned immediately. When the pool is not empty, an object is removed
+ * from the pool and returned. The user is expected to return the object to the pool when it is no
+ * longer used.
+ *
+ * <p>Not thread safe
+ */
+public class ObjectPool<T> {
+  private final ArrayBasedStack<T> pool;
```

Comment:
I don't see any appreciable difference between using the custom `ArrayBasedStack` and the built in `ArrayDeque`

Reply:
There isn't any now. I want to add eviction. 
I want the pool to take a look at the actual usage vs its size. It's capacity might be 1000 but in the last 5 collections, the number of elements never exceeded 550. In that I would like to resize the array to 600.
I won't be able to do that with ArrayDeqeue as the array capacity is internal, hence resizing the array. 

It's important for 2 reasons:
1. The number of attributes can grown but then shrink and stay this way for a while. 
2. As I saw with the `AsynchronousCounterMemoryUsageBenchmark` that we occupy more memory than we should since we grow by a factor every time we need to resize. Eviction + smaller factor (additional tests required) will fix that.